src/esp/sensor/Sensor.cpp

```diff
@@ -51,9 +50,30 @@ Sensor::Sensor(scene::SceneNode& node, SensorSpec::ptr spec)
                  "Sensor::Sensor(): Cannot initialize sensor. The "
                  "specification is null.", );
   spec_->sanityCheck();
+
+  node.getNodeSensorSuite().add(*this);
+  node.getSubtreeSensorSuite().add(*this);
+  // Traverse up to root node and add sensor to every subtreeSensorSuite
+  auto parent = node.parent();
+  while (parent->parent() != nullptr) {
+    static_cast<scene::SceneNode&>(*parent).getSubtreeSensorSuite().add(*this);
+    parent = parent->parent();
+  }
   setTransformationFromSpec();
 }
 
+Sensor::~Sensor() {
```

Comment:
Some problems: When a SceneNode which contains a feature is deleted, it seems that the feature is released so that calling node() segfaults, and we can't remove the reference from relevant SensorSuites.

Reply:
@mosra - could you take a look at this destructor? Specifically, we are having some problems with the order of destructors. We want the destructor for the Sensor to be called before its node is released, but it seems to be the other way around.

Reply:
> We want the destructor for the Sensor to be called before its node is released, but it seems to be the other way around.

It's ... both :) The destruction order for the `SceneNode` is the following:

1. `~SceneNode()` gets called
2. all `SceneNode` member destructors get called
3. `~Object()` gets called
4. `~Object()` iterates through all attached features and `delete`s these
5. `~Object()` iterates through all children and `delete`s these (or maybe this is first and then the features, not sure)

I don't know the backtrace of the crash, but according to what I see done here, I assume it's because at step (4) you're calling `getSubtreeSensorSuite()`, which however gives you a reference to a member that got already destructed at step (2).

Magnum itself can't really do anything about this since at the point the `~Object()` gets called, all subclass state is already gone. 

*[~~30~~ 60 minutes of head scratching pass]*

I see ~~two~~ three options:

- You'd explicitly delete all attached Sensor features in `~SceneNode()`, which basically puts the step (4) before step (2) and it shouldn't crash anymore ~~i think~~ i hope. Not 100% sure in what state the parent nodes would be at that point tho, maybe those are already deleted as well so this only moves the problem one level up ... would need *extensive* tests. Or maybe you go through all children there as well and delete their sensors too? Sounds heavy.
- Or you do a `dynamic_cast<scene::SceneNode*>(current)` and then check the returned pointer. If it's null, then the parent node got already destructed and thus there's no point in cleaning up anything, as all the state is gone already anyway; if it's non-null then the parent is still alive and so it makes sense to do the `remove()`. Though here you need to explicitly test that when you delete just a subtree, the rest of the graph won't be left with outdated information. I think that should be avoided if you continue going up the parent hierarchy even when `dynamic_cast` returns `nullptr` (instead of for example `break`ing after that), but again, that needs an explicit test.
- Or rethink how all this information is managed and what's responsible for updating the tree when operations like deleting a subtree happen. Doesn't seem like a task for one afternoon, though.

@bigbike, this is exactly what i feared of when we were discussing this a few weeks ago, that keeping all this redundant information up-to-date will be a huge and complex task prone to subtle errors :sweat_smile:

Reply:
@mosra: by the way, this is the backtrace:

```
#0  0x000055555572d18d in Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> >::parent (this=0x0)
    at /home/yili/temp/habitat-sim/src/deps/magnum/src/Magnum/SceneGraph/Object.h:151
#1  esp::sensor::Sensor::~Sensor (this=0x5555561d2fe0, __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/esp/sensor/Sensor.cpp:70
#2  0x000055555572c43c in esp::sensor::VisualSensor::~VisualSensor (this=0x5555561d2fe0, 
    __in_chrg=<optimised out>) at /usr/include/c++/9/bits/move.h:99
#3  esp::sensor::CameraSensor::~CameraSensor (this=0x5555561d2fe0, __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/esp/sensor/CameraSensor.h:25
#4  esp::sensor::CameraSensor::~CameraSensor (this=0x5555561d2fe0, __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/esp/sensor/CameraSensor.h:25
#5  0x000055555572c088 in Corrade::Containers::LinkedListItem<Magnum::SceneGraph::AbstractFeature<3u, float>, Magnum::SceneGraph::AbstractObject<3u, float> >::doErase (this=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:345
#6  0x0000555555850633 in Corrade::Containers::LinkedListItem<Magnum::SceneGraph::AbstractFeature<3u, float>, Magnum::SceneGraph::AbstractObject<3u, float> >::erase (this=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:419
#7  Corrade::Containers::LinkedList<Magnum::SceneGraph::AbstractFeature<3u, float> >::erase (
    item=<optimised out>, this=0x5555561d2da8)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:381
#8  Corrade::Containers::LinkedList<Magnum::SceneGraph::AbstractFeature<3u, float> >::clear (
    this=0x5555561d2da8)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:389
#9  Corrade::Containers::LinkedList<Magnum::SceneGraph::AbstractFeature<3u, float> >::~LinkedList (
    this=0x5555561d2da8, __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:123
#10 Magnum::SceneGraph::AbstractObject<3u, float>::~AbstractObject (this=this@entry=0x5555561d2da0, 
    __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/magnum/src/Magnum/SceneGraph/Object.hpp:42
#11 0x0000555555868b68 in Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> >::~Object (this=0x5555561d2da0, __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/EnumSet.h:98
#12 0x000055555573c33b in non-virtual thunk to esp::scene::SceneNode::~SceneNode() ()
    at /home/yili/temp/habitat-sim/src/deps/magnum/src/Magnum/SceneGraph/Object.h:358
#13 0x0000555555728606 in Corrade::Containers::LinkedListItem<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> >, Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::doErase (
    this=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:345
#14 0x0000555555868b10 in Corrade::Containers::LinkedListItem<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> >, Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::erase (this=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:419
#15 Corrade::Containers::LinkedList<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::erase (item=<optimised out>, this=0x5555561d0e68)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:381
#16 Corrade::Containers::LinkedList<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::clear (this=0x5555561d0e68)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:389
#17 Corrade::Containers::LinkedList<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::~LinkedList (this=0x5555561d0e68, 
    __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:123
#18 Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<fl--Type <RET> for more, q to quit, c to continue without paging--
oat> >::~Object (this=0x5555561d0e20, __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/magnum/src/Magnum/SceneGraph/Object.h:105
#19 0x000055555573c33b in non-virtual thunk to esp::scene::SceneNode::~SceneNode() ()
    at /home/yili/temp/habitat-sim/src/deps/magnum/src/Magnum/SceneGraph/Object.h:358
#20 0x0000555555728606 in Corrade::Containers::LinkedListItem<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> >, Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::doErase (
    this=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:345
#21 0x0000555555868b10 in Corrade::Containers::LinkedListItem<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> >, Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::erase (this=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:419
#22 Corrade::Containers::LinkedList<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::erase (item=<optimised out>, this=0x5555561d0bf8)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:381
#23 Corrade::Containers::LinkedList<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::clear (this=0x5555561d0bf8)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:389
#24 Corrade::Containers::LinkedList<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::~LinkedList (this=0x5555561d0bf8, 
    __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:123
#25 Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> >::~Object (this=0x5555561d0bb0, __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/magnum/src/Magnum/SceneGraph/Object.h:105
#26 0x000055555573c33b in non-virtual thunk to esp::scene::SceneNode::~SceneNode() ()
    at /home/yili/temp/habitat-sim/src/deps/magnum/src/Magnum/SceneGraph/Object.h:358
#27 0x0000555555728606 in Corrade::Containers::LinkedListItem<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> >, Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::doErase (
    this=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:345
#28 0x0000555555868b10 in Corrade::Containers::LinkedListItem<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> >, Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::erase (this=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:419
#29 Corrade::Containers::LinkedList<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::erase (item=<optimised out>, this=0x5555561cfa30)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:381
#30 Corrade::Containers::LinkedList<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::clear (this=0x5555561cfa30)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:389
#31 Corrade::Containers::LinkedList<Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> > >::~LinkedList (this=0x5555561cfa30, 
    __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/LinkedList.h:123
#32 Magnum::SceneGraph::Object<Magnum::SceneGraph::BasicTranslationRotationScalingTransformation3D<float> >::~Object (this=0x5555561cf9e8, __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/magnum/src/Magnum/SceneGraph/Object.h:105
#33 0x00005555557278d7 in esp::scene::SceneNode::~SceneNode (this=0x5555561cf9e8, 
    __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/Containers/Optional.h:240
#34 esp::scene::SceneGraph::~SceneGraph (this=0x5555561cf960, __in_chrg=<optimised out>)
--Type <RET> for more, q to quit, c to continue without paging--
    at /home/yili/temp/habitat-sim/src/esp/scene/SceneGraph.h:25
#35 esp::scene::SceneGraph::~SceneGraph (this=0x5555561cf960, __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/esp/scene/SceneGraph.h:25
#36 std::default_delete<esp::scene::SceneGraph>::operator() (this=0x5555561d0b90, 
    __ptr=0x5555561cf960) at /usr/include/c++/9/bits/unique_ptr.h:81
#37 std::unique_ptr<esp::scene::SceneGraph, std::default_delete<esp::scene::SceneGraph> >::~unique_ptr (this=0x5555561d0b90, __in_chrg=<optimised out>) at /usr/include/c++/9/bits/unique_ptr.h:292
#38 std::_Destroy<std::unique_ptr<esp::scene::SceneGraph, std::default_delete<esp::scene::SceneGraph> > > (__pointer=0x5555561d0b90) at /usr/include/c++/9/bits/stl_construct.h:98
#39 std::_Destroy_aux<false>::__destroy<std::unique_ptr<esp::scene::SceneGraph, std::default_delete<esp::scene::SceneGraph> >*> (__last=<optimised out>, __first=0x5555561d0b90)
    at /usr/include/c++/9/bits/stl_construct.h:108
#40 std::_Destroy<std::unique_ptr<esp::scene::SceneGraph, std::default_delete<esp::scene::SceneGraph> >*> (__last=<optimised out>, __first=<optimised out>)
    at /usr/include/c++/9/bits/stl_construct.h:137
#41 std::_Destroy<std::unique_ptr<esp::scene::SceneGraph, std::default_delete<esp::scene::SceneGraph> >*, std::unique_ptr<esp::scene::SceneGraph, std::default_delete<esp::scene::SceneGraph> > > (
    __last=0x5555561d0b98, __first=<optimised out>) at /usr/include/c++/9/bits/stl_construct.h:206

#42 std::vector<std::unique_ptr<esp::scene::SceneGraph, std::default_delete<esp::scene::SceneGraph> >, std::allocator<std::unique_ptr<esp::scene::SceneGraph, std::default_delete<esp::scene::SceneGraph> > > >::~vector (this=0x7fffffffb300, __in_chrg=<optimised out>)
    at /usr/include/c++/9/bits/stl_vector.h:677
#43 esp::scene::SceneManager::~SceneManager (this=0x7fffffffb300, __in_chrg=<optimised out>)
    at /home/yili/temp/habitat-sim/src/esp/scene/SceneManager.h:25
#44 SensorTest::testSensorFactory (this=<optimised out>)
    at /home/yili/temp/habitat-sim/src/tests/SensorTest.cpp:31
#45 0x000055555574cc7d in Corrade::TestSuite::Tester::exec (this=0x7fffffffd9f0, 
    previousTester=<optimised out>, logOutput=<optimised out>, errorOutput=<optimised out>)
    at /home/yili/temp/habitat-sim/src/deps/corrade/src/Corrade/TestSuite/Tester.cpp:476
#46 0x000055555570cda6 in main (argc=<optimised out>, argv=<optimised out>)
    at /home/yili/temp/habitat-sim/src/tests/SensorTest.cpp:115

```

Reply:
Ah. That's because you basically do 

```
node().parent()->parent()
               ^
               *boom*, this might be null aready
```

Reply:
@bigbike @mosra Thank you so much for such an in-depth response!
Yes, I'm finding that node() is actually already null when the destructor is called. This info is really good am I'm trying to see if I can find a workaround.

Reply:
Just for the record after a private discussion with Yili -- the problem was a subtlety in the behavior of LinkedList, which is what the objects and features use to connect to each other -- and should be fixed with https://github.com/mosra/corrade/commit/1daaac6ef5b8aef9eaa31f1def8a8f26c5ca8906, which has https://github.com/facebookresearch/habitat-sim/pull/1123 as a prerequisite.

Reply:
> Thanks for the fix to Magnum! That's works

@vauduong : When you said this, are you sure you applied:
[mosra/corrade@1daaac6](https://github.com/mosra/corrade/commit/1daaac6ef5b8aef9eaa31f1def8a8f26c5ca8906)
Please goto the corrade's source code to check. It is in the src/deps/corrade

>current from auto current = node().parent() will just be nullptr.

@mosra : This worries me a lot. You might be worried too. because the "half-destroyed" theory (SceneNode is destroyed, but BASE is not destroyed.) may not be true (assuming Vivian already applied 1daaac6.